Bluetooth: userchan: Support TCP fragmentation of HCI packets

patstew opened the pull request:
Keep reading from the HCI socket when a packet is incomplete. The other end may not write entire packets, or TCP could fragment in the middle of a packet.
Also fix a potential infinite loop by advancing to the next packet before any continue statements.

jhedberg reviewed (changes requested):
What does Linux HCI user channel sockets have to do with TCP? The Linux kernel should ensure that you always get entire packet frames when reading from a userchan socket. The userchan HCI driver in Zephyr takes advantage of this, and there are also BlueZ tools which assume the same. Could you describe your setup a little more in detail, assuming that you're actually seeing the need for this change in practice?

patstew commented:
That code is also used in native_sim for TCP hci emulation. e.g.

    west build -b native_sim samples/bluetooth/peripheral
    build/zephyr/zephyr.exe --bt-dev=127.0.0.1:9123

jhedberg commented:
@patstew thanks for the clarification. I had completely forgotten that our userchan driver had been extended with TCP support (curiously, even though I was the assignee in the original feature PR, it got merged without my approval). Btw, I think we might want to consider renaming the driver since its name still implies that it's only about user channel sockets.

aescolar commented:
CC @vChavezB

vChavezB commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -155,6 +155,11 @@ static uint16_t packet_len(const uint8_t *buf)
 	}
 
+	/* Prevent overflow (e.g. when we've received a partial header so payload_len is junk) */
+	if (UINT16_MAX - sizeof(type) - header_len < payload_len) {
```
Would it not be better to actually verify if we recieved a partial header? Taking as reference Bluetooth Spec v5.4 Vol 4 Part E you will find that the payload length can be either have a length of 1 byte (H4_CMD, H4_SCO, H4_EVT) or 2 bytes (H4_ACL, H4_ISO) so I am not sure if this work in two edge cases:

   1. When you recieve a fragmented H4 frame that has a payload of maximum 1 byte (i.e., UINT8_MAX)
   2. When you recieve a fragmented H4 frame with just the first payload byte (i.e. H4_ACL, H4_ISO) with a second invalid payload byte that. E.g., Payload len byte 1 = 0x01, Payload len byte 2 (invalid due to fragmentation) = 0xAB. This would bypass your oveflow check and assume you have a payload of size 0x01AB.
   
You could instead use the current size of your frame, and the frame type to verify if the frame payload length is not complete before you actually decode the H4 packet len with `packet_len`.

vChavezB replied:
Example suggestion
```c
static int packet_len(const uint8_t *buf, const uint8_t buf_size, uint16_t * packet_len)
{
	uint16_t payload_len = 0;
	uint8_t header_len = 0;
	const uint8_t type = buf[0];
	const uint8_t *hdr = &buf[sizeof(type)];
	
	if (buf_size ==  sizeof(type)) {
            return -EINVAL;
	}

	switch (type) {
	  case H4_CMD: {
		  const struct bt_hci_cmd_hdr *cmd = (const struct bt_hci_cmd_hdr *)hdr;
  
		  /* Parameter Total Length */
		  payload_len = cmd->param_len;
		  header_len = BT_HCI_CMD_HDR_SIZE;
		  
		  break;
	  }
	    	// rest of the switch cases
	  }

	  if (buf_size  < header_len ) {
	      return -EINVAL
	  }
	  
	  *packet_len = sizeof(type) + header_len + payload_len;
}
```

patstew replied:
Yea I initially did something along those lines, but then thought this way is less code. The way I've done it the length is sometimes "wrong" but is always greater than or equal to the true value. I check it never overflows, and if it has a full header it's correct, otherwise it's the true header length plus some random positive value, so it receives more until it does have a full header. The case where it sends just the initial type byte then the rest of the packet is in fact exactly the situation that motivated this patch.

Happy to change to the above if preferred though.

jhedberg replied:
Something along the lines of what @vChavezB suggested sounds like a better approach to me. That said, we have several examples of H4 decoding in the tree, so might make sense to just follow the same (proven) logic as used there:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/bluetooth/hci/h4.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/hci_uart/src/main.c

patstew replied:
I have changed it to roughly what @vChavezB suggests. I'm not sure it's worth refactoring both files to share with those other bits, because the relevant code is not very complicated and in this specific usage has the unusual requirement to not go wrong if the data is incomplete.

vChavezB commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -114,48 +115,53 @@ static uint16_t packet_len(const uint8_t *buf)
 		/* ISO_Data_Load_Length parameter */
 		payload_len =  bt_iso_hdr_len(sys_le16_to_cpu(iso->len));
-		header_len = BT_HCI_ISO_HDR_SIZE;
+		header_len = BT_HCI_ISO_HDR_SIZE + sizeof(type);
```
This might just be a minor detail but you are repeating the addition of the length of the H4 packet type (i.e., `sizeof(type)`) in all of the cases.  Would it not be better to make the addition after the switch case? It will not affect the logic but improves readability and removes code repetition.

vChavezB commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -99,9 +99,10 @@ static struct net_buf *get_rx(const uint8_t *buf)
  * @details Decodes packet length according to Bluetooth spec v5.4 Vol 4 Part E
  * @param buf	Pointer to a HCI packet buffer
- * @return Length of the HCI packet in bytes, zero if no valid packet found.
+ * @return Length of the HCI packet in bytes, -1 if no valid packet found, 0 if
```
Originally the purpose of this function was to only decode the packet length format. With the new logic introduce, this function will not decode the packet length if the frame is not complete. Perhaps this should be mentioned for mantainance/debugging purposes.

vChavezB commented:
Just added two comments on the changes. They are minor details but hope they help. Thanks for the contribution 👍  I am happy to know that other users are taking advantage of connecting Zephyr applications to other external controllers.

patstew committed:
Bluetooth: userchan: Support TCP fragmentation of HCI packets

Keep reading from the HCI socket when a packet is incomplete. The other
end may not write entire packets, or TCP could fragment in the middle of a
packet.
Also fix a potential infinite loop by advancing to the next packet before
any continue statements.

Signed-off-by: Patrick Stewart <patrick@rfcreations.com>

vChavezB commented:
Looks good to me, I think you just need to await feedback from the code owners or collaborators of the project.

patstew commented:
I'd like the stale label removed (and for someone else to review it)
